### PR TITLE
Streamline latest block access & finalized block caching

### PR DESCRIPTION
**Pull request by nounderline**
- Cache finalized block fetching globally. Before `getBlock` RPC call was made for every `getAccountHistory` request. Now the value is refreshed at most once every 1s. More: #1180
- Fix a bug where `AccountHistoryResult` could be one block behind. Now we get latest indexed block from Shovel database instead from internal `Watcher` indexer in `getAccountHistory`. This bug occurred when event loop called a new block notification callback in `Watcher.watch` before one in `onAccountUpdate` TRPC subscription.

### CODE REVIEW

File: packages/daimo-api/src/server/router.ts [hunk 1]
```diff
@@ -756,42 +756,42 @@ export function createRouter(
...
-
-              emit.next(history);
-
-              lastEmittedBlock = history.lastBlock;
-            });
+              extApiCache,
+              blockNumber
+            );
+
+            if (lastEmittedBlock > history.lastBlock) {
```
Comment (dcposch) [inline packages/daimo-api/src/server/router.ts:763]: should this be `>=` ?

File: packages/daimo-api/src/utils/cache.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+ *
+ * `cacheTime` is duration in ms after cached value is considred expired,
+ * and must be refreshed before it can be returned.
+ *
+ * `staleTime` is an optional duration that refreshes
+ * the cache before it expires. Use it to avoid blocking after cache expires.
+ * Must be smaller than `cacheTime`.
+ *
+ */
+export function lazyCache<T>(
```
Comment (dcposch) [inline packages/daimo-api/src/utils/cache.ts:14]: this is interesting. feels like a lot of complexity for just `getFinalizedBlock`

finalizedBlock changes slowly & doesn't affect UI latency. is the goal to reduce JSON RPC API calls?

**Review by dcposch — APPROVED** [review]
Left comments, LGTM otherwise

File: packages/daimo-api/src/utils/cache.ts (hunk 2 above)
Comment (nounderline) [reply packages/daimo-api/src/utils/cache.ts:14]: Finalized block is done on every account history fetch. It was [second slowest operation](https://github.com/daimo-eth/daimo/issues/1180#issuecomment-2186485047) in `getAccountHistory`. Now we share finalized block data across all users limiting RPC calls from variable `n` to constant `2` every ~2 sec, where `n` is number of connected users.

**Pull request merged**